uniffi_bindgen/src/bindings/kotlin/templates/ErrorTemplate.kt

```diff
@@ -1,149 +1,102 @@
+        {% endfor -%}
+    ) : {{ toplevel_name }}()
+    {%- endif %}
+    {% endfor %}
+
+    companion object ErrorHandler : CallStatusErrorHandler<{{ toplevel_name }}> {
+        override fun lift(error_buf: RustBuffer.ByValue): {{ toplevel_name }} {
+            return liftFromRustBuffer(error_buf) { error_buf -> read(error_buf) }
         }
-        val message = this.consumeErrorMessage()
-        when (code) {
-            {% for value in e.values() -%}
-            {{loop.index}} -> return {{e|exception_name_kt}}.{{value}}(message) as E
-            {% endfor -%}
-            -1 -> return InternalException(message) as E
-            else -> throw RuntimeException("invalid error code passed across the FFI")
+
+        fun read(error_buf: ByteBuffer): {{ toplevel_name }} {
+            return when(error_buf.getInt()) {
+                {%- for variant in e.variants() %}
+                {{ loop.index }} -> {{ toplevel_name }}.{{ variant.name()|exception_name_kt }}({% if variant.has_fields() %}
+                    {% for field in variant.fields() -%}
+                    {{ "error_buf"|read_kt(field.type_()) }}{% if loop.last %}{% else %},{% endif %}
+                    {% endfor -%}
+                {%- endif -%})
+                {%- endfor %}
+                else -> throw RuntimeException("invalid error enum value, something is very wrong!!")
+            }
         }
     }
 }
-
-open class {{e|exception_name_kt}}(message: String) : Exception(message) {
-    {% for value in e.values() -%}
-    class {{value}}(msg: String) : {{e|exception_name_kt}}(msg)
-    {% endfor %}
-}
-
 {% endfor %}
 
-// Helpers for calling Rust with errors:
+// Helpers for calling Rust
 // In practice we usually need to be synchronized to call this safely, so it doesn't
 // synchronize itself
-private inline fun <U, E: RustErrorReference> nullableRustCall(callback: (E) -> U?, err: E): U? {
-    try {
-        val ret = callback(err)
-        if (err.isFailure()) {
-            throw err.intoException()
+
+// Call a rust function that returns a Result<>.  Pass in the Error class companion that corresponds to the Err
+private inline fun <U, E: Exception> rustCallWithError(errorHandler: CallStatusErrorHandler<E>, callback: (RustCallStatus) -> U): U {
+    var status = RustCallStatus();
+    val return_value = callback(status)
+    if (status.isSuccess()) {
+        return return_value
+    } else if (status.isError()) {
+        throw errorHandler.lift(status.error_buf)
+    } else if (status.isPanic()) {
+        // when the rust code sees a panic, it tries to construct a rustbuffer
+        // with the message.  but if that code panics, then it just sends back
+        // an empty buffer.
+        if (status.error_buf.len > 0) {
+            throw InternalException(String.lift(status.error_buf))
+        } else {
+            throw InternalException("Rust panic")
```

Comment:
Do we need to free `error_buf` here?

Reply:
I think we can avoid freeing since it's the same `RustBuffer` that was initially passed in with `data=null`.